### ISSUE TITLE
Install-app-deps Fails with SQLite3 in React and Electron Application

### ISSUE THREAD

**Issue by StatelierVirus4** [electron-userland/electron-builder#4658]
* **Electron-Builder Version**: 22.3.2
* **Electron**: 8.0.0
* **npm Version**: 6.13.6
* **nodejs Version**: 13.8.0
* Latest version of Linux Mint

I'm trying to create a React/Electron application that uses sqlite3 as a dependency.

I used the following commands to build my project:
```
npx create-react-app testproject
npm install electron
npm install -D electron-builder
npm install sqlite3
```
When I run `npm run [electron-builder install-app-deps]` I get the following error:

```
• electron-builder  version=22.3.2
  • public/electron.js not found. Please see https://medium.com/@kitze/%EF%B8%8F-from-react-to-an-electron-app-ready-for-production-a0468ecb1da3
  • loaded parent configuration  preset=react-cra
  • rebuilding native dependencies  dependencies=sqlite3@4.1.1 platform=linux arch=x64
  ⨯ cannot execute  cause=exit status 1
                    out=
    > sqlite3@4.1.1 install /home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3
    > node-pre-gyp install --fallback-to-build
    
    make: Entering directory '/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3/build'
      ACTION deps_sqlite3_gyp_action_before_build_target_unpack_sqlite_dep Release/obj/gen/sqlite-autoconf-3300100/sqlite3.c
      TOUCH Release/obj.target/deps/action_before_build.stamp
      CC(target) Release/obj.target/sqlite3/gen/sqlite-autoconf-3300100/sqlite3.o
    deps/sqlite3.target.mk:139: recipe for target 'Release/obj.target/sqlite3/gen/sqlite-autoconf-3300100/sqlite3.o' failed
    make: Leaving directory '/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3/build'
    Failed to execute '/usr/bin/node /usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3/lib/binding/electron-v8.0-linux-x64/node_sqlite3.node --module_name=node_sqlite3 --module_path=/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3/lib/binding/electron-v8.0-linux-x64 --napi_version=5 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=electron-v8.0' (1)
    
                    errorOut=node-pre-gyp WARN Using request for node-pre-gyp https download 
    node-pre-gyp WARN Tried to download(403): https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v4.1.1/electron-v8.0-linux-x64.tar.gz 
    node-pre-gyp WARN Pre-built binaries not found for sqlite3@4.1.1 and electron@8.0.0 (electron-v8.0 ABI, glibc) (falling back to source compile with node-gyp) 
    Release/obj/gen/sqlite-autoconf-3300100/sqlite3.c:13763:10: fatal error: stdio.h: No such file or directory
     #include <stdio.h>
              ^~~~~~~~~
    compilation terminated.
    make: *** [Release/obj.target/sqlite3/gen/sqlite-autoconf-3300100/sqlite3.o] Error 1
    gyp ERR! build error 
    gyp ERR! stack Error: `make` failed with exit code: 2
    gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
    gyp ERR! stack     at ChildProcess.emit (events.js:321:20)
    gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
    gyp ERR! System Linux 4.15.0-54-generic
    gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "build" "--fallback-to-build" "--module=/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3/lib/binding/electron-v8.0-linux-x64/node_sqlite3.node" "--module_name=node_sqlite3" "--module_path=/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3/lib/binding/electron-v8.0-linux-x64" "--napi_version=5" "--node_abi_napi=napi" "--napi_build_version=0" "--node_napi_label=electron-v8.0"
    gyp ERR! cwd /home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3
    gyp ERR! node -v v13.8.0
    gyp ERR! node-gyp -v v5.0.5
    gyp ERR! not ok 
    node-pre-gyp ERR! build error 
    node-pre-gyp ERR! stack Error: Failed to execute '/usr/bin/node /usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js build --fallback-to-build --module=/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3/lib/binding/electron-v8.0-linux-x64/node_sqlite3.node --module_name=node_sqlite3 --module_path=/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3/lib/binding/electron-v8.0-linux-x64 --napi_version=5 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=electron-v8.0' (1)
    node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/node-pre-gyp/lib/util/compile.js:83:29)
    node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:321:20)
    node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:1026:16)
    node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)
    node-pre-gyp ERR! System Linux 4.15.0-54-generic
    node-pre-gyp ERR! command "/usr/bin/node" "/home/linuxmint/Documents/ElectronApplications/testproject/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
    node-pre-gyp ERR! cwd /home/linuxmint/Documents/ElectronApplications/testproject/node_modules/sqlite3
    node-pre-gyp ERR! node -v v13.8.0
    node-pre-gyp ERR! node-pre-gyp -v v0.11.0
    node-pre-gyp ERR! not ok 
    npm ERR! code ELIFECYCLE
    npm ERR! errno 1
    npm ERR! sqlite3@4.1.1 install: `node-pre-gyp install --fallback-to-build`
    npm ERR! Exit status 1
    npm ERR! 
    npm ERR! Failed at the sqlite3@4.1.1 install script.
    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/linuxmint/.npm/_logs/2020-02-08T20_47_34_468Z-debug.log
    
                    command=/usr/bin/node /usr/lib/node_modules/npm/bin/npm-cli.js rebuild sqlite3@4.1.1
                    workingDir=
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! testproject@0.1.0 installsqlitescript: `electron-builder install-app-deps`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the testproject@0.1.0 installsqlitescript script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/linuxmint/.npm/_logs/2020-02-08T20_47_34_497Z-debug.log

```
Adding public/electron.js doesn't help.

I've found various solutions to problems that have similar errors to the one that I'm getting, none of which have worked for me.

One thing that I haven't tried is downgrading some of the different software packages that the project uses, which I don't want to do.

I'm not sure what I'm doing wrong. Any help is appreciated!

**Comment by ThranorBeardhammer** [comment]
I am also having an issue getting sqlite3 installed.  This line seems to be the culprit?

> node-pre-gyp WARN Tried to download(403): https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v4.1.1/electron-v8.0-linux-x64.tar.gz

**Comment by StatelierVirus4** [comment]
Thanks for pointing that out! This is similar to the problem here: 
https://github.com/mapbox/node-sqlite3/issues/918

I went ahead and downgraded to the most recent version of electron 7 inside my package.json, which was **7.1.12**. I ran `npm install` to update the version inside my project. When I ran `npm run [electron-builder install-app-deps]`, it worked fine and I was able to run my electron application with `const sqlite3 = require("sqlite3")` included in my main.js. 

I'm guessing this means that there isn't a binary available for node-pre-gyp to use for version 8.0.0 of electron yet?

**Comment by ThranorBeardhammer** [comment]
Perfect!  worked for me too.   I wonder if there is a way to keep updated on what version are available to see when v8 has a version out.

**Comment by StatelierVirus4** [comment]
It has been a while but I was still having trouble a few months after this post and I finally got around to posting what I think what the problem was for me.

I didn't see on the SQLite3 GitHub page that they have a list of Electron versions that that SQLite3 works with. Since my Electron version didn't match any of those versions, I needed to downgrade to one of those versions.

However, on the SQLite3 page, it states:

> "The module uses node-pre-gyp to download a pre-compiled binary for your platform, if it exists. Otherwise, it uses node-gyp to build the extension."

But it wasn't building the binaries for my version of Electron.

When I checked the node-gyp page, I noticed that it said for the requirements that I needed

> "A proper C/C++ compiler toolchain, like GCC"

and
> "make"

After installing both of those, uninstalling SQLite3 (and deleting the node_modules and package-lock.json files), running `npm install sqlite3`, running `npm run electron-builder install-app-deps`, I was good to go!

However, I believe this causes issues when trying to package the application for other platforms. In my case, the binaries are built in Linux Mint but I wanted to build for Windows 10. The docker solution in the electron-builder didn't work for me because I think SQLite3 is now a native dependency since I had to build binaries. So I needed to use AppVeyor to build for Windows.
